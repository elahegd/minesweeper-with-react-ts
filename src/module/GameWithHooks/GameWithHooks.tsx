import { FC, ChangeEvent, useCallback } from 'react';

import { GameArea } from '../../components/Game/GameArea';
import { GameOver } from '../../components/Game/GameOver';
import { Grid } from "../../components/Grid/Grid";
import { Top } from "../../components/Top";
import { Scoreboard } from '../../components/Scoreboard';

import { GameLevel, LevelNames } from "../../module/GameSettings";

import { useGame } from "../../hooks/useGame";

export const GameWithHooks:FC = () => {
  const { 
    level,
    isGameOver,
    isWin,
    playerField,
    setting,
    time,
    flagCounter,
    handleClick,
    handleChangeLevel,
    onReset,
    handleContextMenu
  } = useGame();

  const [, bombs] = setting;

  const handleChange = useCallback(
    ({target: {value: level}}: ChangeEvent<HTMLSelectElement>) => 
      handleChangeLevel(level as LevelNames), 
      // Stryker disable next-line ArrayDeclaration
      []);

  return (
    <>
        <Top feature="Flag" firstAction="ctrl" secondAction="click">Minesweeper</Top>
        <GameArea>
          <Scoreboard 
            timer={String(time)}
            levels={GameLevel as unknown as string[]}
            defaultValue={level}
            onChangeLevel={handleChange}
            onReset={onReset}
            bombs={String(bombs - flagCounter)}
          />
          {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
          <Grid onClick={handleClick} onContextMenu={handleContextMenu}>
              {playerField}
          </Grid>
        </GameArea>
    </>
  )
}