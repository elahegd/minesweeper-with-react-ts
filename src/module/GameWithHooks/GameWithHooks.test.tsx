import React from 'react';
import { useSearchParams } from 'react-router-dom';
import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";

import { GameWithHooks } from "./GameWithHooks";

const mockHandleClick = jest.fn();
const mockHandleChangeLevel = jest.fn();
const mockOnReset = jest.fn();
const mockHandleContextMenu = jest.fn();
const mockSetSearchParams = jest.fn();


const searchParams = { get: () => null };
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as object),
    useSearchParams: () => [searchParams, mockSetSearchParams]
}));


jest.mock('../../hooks/useGame', () => ({
    __esModule: true,
    useGame: (level = 'beginner') => ({
        level,
        time: 0,
        isGameOver: true,
        isWin: false,
        flagCounter: 0,
        setting: [9, 10],
        playerField: [
            [10, 10],
            [10, 10]
        ],
        handleClick: mockHandleClick,
        handleChangeLevel: mockHandleChangeLevel,
        onReset: mockOnReset,
        handleContextMenu: mockHandleContextMenu
    })
}));

beforeEach(() => {
    jest.clearAllMocks();
});

describe("Game with hooks test cases", () => {
    it("Render game field by default", () => {
        const { asFragment } = render(<GameWithHooks />);
        expect(asFragment()).toMatchSnapshot();
    });
    it("Cell click works fine", () => {
        render(<GameWithHooks />);
        userEvent.click(screen.getByTestId("0,0"));
        expect(mockHandleClick).toHaveBeenCalled();
    });
    it("Reset handler works fine", () => {
        render(<GameWithHooks />);
        userEvent.click(screen.getByRole('button'));
        expect(mockOnReset).toHaveBeenCalled();
    });
    it("Change level works fine", () => {
        render(<GameWithHooks />);
        userEvent.selectOptions(screen.getByRole('combobox'), 'intermediate');
        expect(mockHandleChangeLevel).toHaveBeenCalled();
        expect(mockSetSearchParams).toHaveBeenCalledWith({
            level: 'intermediate'
        })
    });
    it('Level in search params works fine', () => {
        // mock the first argument
        jest.mock('react-router-dom', () => ({
            ...(jest.requireActual('react-router-dom') as object),
            useSearchParams: () => [{get: () => 'intermediate'}, mockSetSearchParams]
        }));
        render(<GameWithHooks />);
        const intermediateOption = screen.queryByText('intermediate');
        expect(intermediateOption).toBeInTheDocument();
      });

    it("Game over resets the game state", () => {
        render(<GameWithHooks />);
        userEvent.click(screen.getByText('🙁'));
        expect(mockOnReset).toHaveBeenCalled();
    });
    it("Context menu handler", () => {
        render(<GameWithHooks />);
        // {button:2} => press the secondbutton of the mouse
        userEvent.click(screen.getByTestId('0,0'), { button: 2});
        expect(mockHandleContextMenu).toHaveBeenCalled();
    });
});