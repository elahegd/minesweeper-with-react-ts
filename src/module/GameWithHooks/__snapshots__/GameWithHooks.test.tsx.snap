// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game with hooks test cases Render game field by default 1`] = `
<DocumentFragment>
  <header
    class="css-dn2347"
  >
    <h1
      class="css-18kvaxm"
    >
      Minesweeper
    </h1>
    <legend
      class="css-17qz6a2"
    >
      <strong>
        Flag: 
      </strong>
      <code
        class="css-12ap3z0"
      >
        <span
          class="css-udfepp"
        >
          ctrl
        </span>
        <span
          class="css-9c23k1"
        >
           + click
        </span>
      </code>
    </legend>
  </header>
  <div
    class="css-hajvvz"
  >
    <div
      class="css-1n7y7g1"
    >
      <div
        class="css-d2q1pa"
      >
        0
      </div>
      <div
        class="css-ov1ktg"
      >
        <select
          class="css-1lpmfg0"
        >
          <option
            class="css-zcmdxr"
            value="beginner"
          >
            beginner
          </option>
          <option
            class="css-zcmdxr"
            value="intermediate"
          >
            intermediate
          </option>
          <option
            class="css-zcmdxr"
            value="expert"
          >
            expert
          </option>
        </select>
        <div>
          <button
            class="css-1kb3z2h"
          >
            😀
          </button>
        </div>
      </div>
      <div
        class="css-d2q1pa"
      >
        10
      </div>
    </div>
    <div
      class="css-ap4m5y"
    >
      🙁
    </div>
    <div
      class="css-17280v7"
      size="2"
    >
      <div
        class="css-1tt7j02"
        data-testid="0,0"
        role="cell"
      >
        10
      </div>
      <div
        class="css-1tt7j02"
        data-testid="0,1"
        role="cell"
      >
        10
      </div>
      <div
        class="css-1tt7j02"
        data-testid="1,0"
        role="cell"
      >
        10
      </div>
      <div
        class="css-1tt7j02"
        data-testid="1,1"
        role="cell"
      >
        10
      </div>
    </div>
  </div>
</DocumentFragment>
`;
